### PR TITLE
backend-data: add support for first-class defineFunction

### PR DESCRIPTION
**Pull request by iartemiev**
## Problem

Allows inline `defineFunction` in the schema builder without additional configuration in `defineData`
Complement to this data-schema PR: https://github.com/aws-amplify/amplify-api-next/pull/124

Docs PR: https://github.com/aws-amplify/docs/pull/6836/files#diff-dfe0a992e880882948237e578edffdfb4a4548f19487a58118ff900d2c7c233aR146-R161

**Issue number, if available:**
fixes: https://github.com/aws-amplify/amplify-backend/issues/573

## Changes

Enables the following DX:
```ts
import {
  type ClientSchema,
  a,
  defineData,
  defineFunction,
} from '@aws-amplify/backend';

const fn1 = defineFunction({entry: './echo/handler.ts',});
const fn2 = defineFunction({entry: './echo/handler2.ts',});

const s = a.schema({
  getPostDetails: a
    .query()
    .arguments({ postId: a.string() })
    .handler([
      a.handler.function(fn1),
      a.handler.function(fn2),
    ])
    .authorization([a.allow.private()])
    .returns(a.string()),
});

export type Schema = ClientSchema<typeof schema>;

export const data = defineData({
  schema
});
```

## Validation
* e2e test (updated)

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [x] If this PR requires a docs update, I have linked to that docs PR above.
- [x] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts [hunk 1]
```diff
@@ -38,7 +38,14 @@ const schema = a.schema({
     .arguments({ content: a.string() })
     .returns(a.ref('EchoResponse'))
     .authorization([a.allow.private()])
-    .function('echo'),
+    .handler(
```
Comment (edwardfoyle) [inline packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:41]: is the old way still supported? If so, maybe we should add this as another resolver and keep the existing one

**Review by edwardfoyle — DISMISSED** [review]
Just a small testing suggestion / clarification

File: packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts (hunk 1 above)
Comment (iartemiev) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:41]: It's marked as deprecated now and will be removed before GA. 

However, you can still pass a string to the new handler, e.g. 
```ts
getPostDetails: a
    .query()
    .arguments({ postId: a.string() })
    .handler([
      a.handler.function('myFunc'),
    ])
```
In which case you'll have to configure via `DataProps.functions`

Comment (iartemiev) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:41]: We can test both here

**Review by sobolk — APPROVED** [review]

File: packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:41]: Is there a use case where you would have to specify the name and the separate mapping? Maybe we remove the `functions` prop entirely from `defineData` and just allow passing the references through the schema?

Comment (iartemiev) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:41]: Updated

**Review by sobolk — APPROVED** [review]

File: packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts (hunk 1 above)
Comment (iartemiev) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:48]: I don't think it's relevant to Amplify Gen2 specifically. We kept the string param in place as an escape hatch for the Amplify CDK construct

Comment (edwardfoyle) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:48]: In that case I'd be in favor of removing the `functions` map entirely. Any objections @sobolk?

Comment (sobolk) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:48]: None. less is better.

Comment (edwardfoyle) [reply packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/resource.ts:48]: After offline discussion, we decided to wait and make this change in a separate PR

**Review by edwardfoyle — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

**Pull request merged**